### PR TITLE
Support passing values into `dbg` with the pipe operator

### PR DESCRIPTION
**Pull request by mulias**
This PR (hopefully) wraps up the work started in #7038 and continued in #7054, fully implementing https://github.com/roc-lang/roc/issues/5894, `dbg` as an expression. 

In the feedback I got for #7038 there was general agreement that the functions in `desugar.rs` could use some refactoring to consolidate params into a shared struct. My initial intuition was that this should be a new struct, independent of the existing `can::Env` struct. Once I started implementing the desugaring for `|> dbg` I realized I would have to pass desugar specific args through the rest of canonicalization, which means the two parts essentially need the same set of args and it makes more sense to share one struct.

### Use a shared env for desugaring and the rest of canonicalization

This refactor simplifies the desugar pass by reducing the number of arguments threaded through each recursive function call.

- Add the module src string to `Env`.
- Add `line_info` to `Env` as a lazy-evaled getter function.
- Refactor desugar functions to take the `can::Env` struct in place of a number of params. This is mostly a find-and-replace, but in a few places `Vec::from_iter_in` was changed to `Vec::with_capacity_in` followed by a `for` loop in order to avoid lifetime issues.
- Remove unnecessary linter annotations for `clippy::too_many_arguments`

### Support passing values into dbg with the pipe operator

In order to desugar `dbg` in a pipeline we need to allow a bare `dbg` node in desugaring and only report it as an error if the bare node survives to the next step of canonicalization. This means we move the error code out of `desugar_expr` and into `canonicalize_expr`. This is much simpler to do now that these functions use the same `env` struct, since previously we would have had to pass down extra args to `canonicalize_expr`. Sharing the `env` struct means that we also don't have to worry about calculating `line_info` more than once.

### CODE REVIEW

File: crates/compiler/can/tests/snapshots/test_suffixed__suffixed_tests__pizza_dbg.snap [hunk 1]
```diff
@@ -1,14 +1,14 @@
 ---
 source: crates/compiler/can/tests/test_suffixed.rs
-assertion_line: 459
+assertion_line: 473
```
Comment (mulias) [inline crates/compiler/can/tests/snapshots/test_suffixed__suffixed_tests__pizza_dbg.snap:3]: This file was accidentally committed early as part of a previous PR. I've updated the source code for the test, so the contents of the snapshot have updated, even though it looks like a new test case.

**Review by smores56 — APPROVED** [review]
I tested this out with basic-cli v0.15, all of my `dbg` usages were working! My past self would have really appreciated this last year, you've really improved the quality of life of Roc debugging here.

**Comment by smores56** [comment]
This is ready to merge, just waiting for whether @mulias agrees with my PR comments or not. Once that's resolved, we're good to go

File: crates/compiler/can/src/desugar.rs [hunk 2]
```diff
@@ -1837,84 +1306,71 @@ fn desugar_dbg_expr<'a>(
...
-        module_path,
-    )
+    desugar_dbg_expr(env, scope, placeholder_expr, outer_region)
 }
 
 /// Desugars a `dbg x` statement into essentially `Inspect.toStr x |> LowLevelDbg`
 fn desugar_dbg_stmt<'a>(
-    arena: &'a Bump,
+    env: &mut Env<'a>,
+    _scope: &mut Scope,
```
[the remark this answers is not in the document]
Comment (mulias) [reply crates/compiler/can/src/desugar.rs:1335]: good catch

**Comment by mulias** [comment]
@smores56 I added two commits to address your feedback. I think the doc comment is more helpful where it is now, but let me know what you think.

**Comment by smores56** [comment]
Uh oh, it looks like you didn't propagate the function arg change...

```
error[E0061]: this function takes 3 arguments but 4 arguments were supplied
    --> crates/compiler/can/src/desugar.rs:1310:17
     |
1310 |         value: *desugar_dbg_stmt(env, scope, tmp_var, tmp_var),
     |                 ^^^^^^^^^^^^^^^^    -------
     |                                     | |
     |                                     | unexpected argument of type `&mut scope::Scope`
     |                                     help: remove the extra argument
     |
note: function defined here
    --> crates/compiler/can/src/desugar.rs:1333:4
     |
1333 | fn desugar_dbg_stmt<'a>(
     |    ^^^^^^^^^^^^^^^^
1334 |     env: &mut Env<'a>,
     |     -----------------
1335 |     condition: &'a Loc<Expr<'a>>,
     |     ----------------------------
1336 |     continuation: &'a Loc<Expr<'a>>,
     |     -------------------------------
```

**Comment by mulias** [comment]
Never pays to rush a change!

**Pull request merged**